kafka-client/src/main/java/dev/responsive/db/CassandraIdempotentKeyValueSchema.java

```diff
@@ -0,0 +1,289 @@
+import static dev.responsive.db.ColumnNames.OFFSET;
+
+import com.datastax.oss.driver.api.core.cql.BoundStatement;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.Row;
+import com.datastax.oss.driver.api.core.type.DataTypes;
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder;
+import com.datastax.oss.driver.api.querybuilder.SchemaBuilder;
+import dev.responsive.db.partitioning.SubPartitioner;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.CheckReturnValue;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CassandraIdempotentKeyValueSchema implements RemoteKeyValueSchema {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      CassandraIdempotentKeyValueSchema.class);
+
+  private final CassandraClient client;
+
+  // use ConcurrentHashMap instead of ConcurrentMap in the declaration here
+  // because ConcurrentHashMap guarantees that the supplied function for
+  // computeIfAbsent is invoked exactly once per invocation of the method
+  // if the key is absent, else not at all. this guarantee is not present
+  // in all implementations of ConcurrentMap
+  private final ConcurrentHashMap<String, PreparedStatement> get;
+  private final ConcurrentHashMap<String, PreparedStatement> insert;
+  private final ConcurrentHashMap<String, PreparedStatement> delete;
+  private final ConcurrentHashMap<String, PreparedStatement> getMeta;
+  private final ConcurrentHashMap<String, PreparedStatement> setOffset;
+
+  public CassandraIdempotentKeyValueSchema(final CassandraClient client) {
+    this.client = client;
+    get = new ConcurrentHashMap<>();
+    insert = new ConcurrentHashMap<>();
+    delete = new ConcurrentHashMap<>();
+    getMeta = new ConcurrentHashMap<>();
+    setOffset = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void create(final String tableName) {
+    LOG.info("Creating data table {} in remote store.", tableName);
+    client.execute(SchemaBuilder
+        .createTable(tableName)
+        .ifNotExists()
+        .withPartitionKey(DATA_KEY.column(), DataTypes.BLOB)
+        .withColumn(DATA_VALUE.column(), DataTypes.BLOB)
+        .withColumn(OFFSET.column(), DataTypes.BIGINT)
+        .build()
+    );
+  }
+
+  /**
+   * Initializes the metadata entry for {@code table} by adding a
+   * row with key {@code _metadata} and sets special columns
+   * {@link ColumnNames#OFFSET} and no {@link ColumnNames#EPOCH}.
+   *
+   * <p>Note that this method is idempotent as it uses Cassandra's
+   * {@code IF NOT EXISTS} functionality.
+   *
+   * @param table          the table that is initialized
+   * @param kafkaPartition the partition to initialize
+   */
+  @Override
+  public FencingToken init(
```

Comment:
The problem makes sense to me, I will change this in a follow-up PR because I think it would be every better to have the fencing token (I'll rename it) be a factory for the atomic writer. That way we don't have to batch at all if we don't want to for the fact table schema.